src/wikigraph.cc

```diff
@@ -70,13 +73,61 @@ std::vector<std::string> WikiGraph::getPathBFS(const std::string& from_page, con
 }
 
 // TODO: getPathDijkstras()
-std::vector<std::string> WikiGraph::getPathDijkstras(const std::string& from_page, const std::string& to_page) const {
-  std::vector<std::string> path_to_return;
-  // ...
-  (void) from_page;
-  (void) to_page;
+std::vector<std::string> WikiGraph::getPathDijkstras(const std::string& start_page, const std::string& end_page) const {
+
+// extract keys from map
+std::vector<std::string> pages;
+std::transform(article_map.begin(), article_map.end(), std::back_inserter(pages),
+    [](decltype(article_map)::value_type const &kv) {
+        return kv.first;
+    });
+
+ // init distance and predecessor
+ std::map<std::string, int> distance;
+ std::map<std::string, std::string> predecessor;
+ for (const auto& page: pages) {
+  distance[page] = INT_MAX;
+  predecessor[page] = "";
+ }
+ distance[start_page] = 0;
+
+ // priority is minimum distance, represented by this lambda expression
+ auto priority = [&distance](std::string a, std::string b) {
+    return distance.at(a) > distance.at(b);
+  };
+
+ // construct the priority queue based on the pages (keys in article_map)
+ std::priority_queue<std::string, std::vector<std::string>, decltype(priority)> q{priority};
+ q.push(start_page);
+
+  // go through the edges
+  while (!q.empty()) {
+    std::string curr = q.top();
+    q.pop();
+    for (const auto& neighbor : article_map.at(curr)) {
+      if (1 + distance[curr] < distance[neighbor]) {
+        distance[neighbor] = 1 + distance[curr];
+        predecessor[neighbor] = curr;
+        q.push(neighbor);
+      }
+    }
+  }
 
-  return path_to_return;
+  // construct the path using the predecessor map
+  std::vector<std::string> path;
+  path.push_back(end_page);
+
+  while (path.back() != start_page && predecessor[path.back()] != "") {
+    path.push_back(predecessor[path.back()]);
+  }
+
+  // there is no path from the start to the end
+  if (std::find(path.begin(), path.end(), start_page) == path.end()) {
+    throw std::invalid_argument("There is no path between these articles");
+  }
+  
+  // return the reversed version (start -> end)
+  return {path.rbegin(), path.rend()};
```

Comment:
This reverse one liner is great! We'll have to put this in the BFS Implementation. I have a lot of wasted lines just reversing the path.